src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.InvalidCast.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class CustomConverterTests
+    {
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/41146")]
+        public static void InvalidCastRefTypedPropertyFails()
+        {
+            var obj = new ObjectWrapperWithProperty
+            {
+                Object = new WrittenObject
+                {
+                    Int = 123,
+                    String = "Hello",
+                }
+            };
+
+            var json = JsonSerializer.Serialize(obj);
+
+            var ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ObjectWrapperWithProperty>(json));
```

Comment:
In 3.1 we throw InvalidCastException; that seems more appropriate than JsonException which typically means the JSON is invalid for the given object model, not that a converter did the wrong thing.

Same applies to other tests here.

Reply:
These tests are annotated with `ActiveIssue` in this PR. The unique exception type (JsonException or another new exception type) was introduced because 
@layomia:

> I think we should create and throw a new exception type explaining that the value returned by the converter is invalid for the property type.